diff -ruN -U3 a/src/bundle_signing.rs b/src/bundle_signing.rs
--- a/src/bundle_signing.rs	2025-11-23 15:42:04
+++ b/src/bundle_signing.rs	2025-11-23 15:59:04
@@ -469,6 +466,14 @@
     ) -> Result<(PathBuf, SignedMachOInfo), AppleCodesignError> {
         warn!("signing Mach-O file {}", bundle_rel_path.display());
 
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::PermissionsExt;
+            let mut perms = std::fs::metadata(source_path)?.permissions();
+            perms.set_mode(0o755);
+            std::fs::set_permissions(source_path, perms)?;
+        }
+
         let macho_data = std::fs::read(source_path)?;
         let signer = MachOSigner::new(&macho_data)?;
 
@@ -607,11 +612,6 @@
             dest_dir.join("Contents")
         };
 
-        self.bundle
-            .identifier()
-            .map_err(AppleCodesignError::DirectoryBundle)?
-            .ok_or_else(|| AppleCodesignError::BundleNoIdentifier(self.bundle.info_plist_path()))?;
-
         let mut resources_digests = settings.all_digests(SettingsScope::Main);
 
         // State in the main executable can influence signing settings of the bundle. So examine
@@ -722,6 +722,14 @@
         // Seal the main executable.
         if let Some(exe) = main_exe {
             warn!("signing main executable {}", exe.relative_path().display());
+
+            #[cfg(unix)]
+            {
+                use std::os::unix::fs::PermissionsExt;
+                let mut perms = std::fs::metadata(exe.absolute_path())?.permissions();
+                perms.set_mode(0o755);
+                std::fs::set_permissions(exe.absolute_path(), perms)?;
+            }
 
             let macho_data = std::fs::read(exe.absolute_path())?;
             let signer = MachOSigner::new(&macho_data)?;
diff -ruN -U3 a/src/lib.rs b/src/lib.rs
--- a/src/lib.rs	2025-12-09 17:09:33
+++ b/src/lib.rs	2025-12-09 17:06:18
@@ -143,7 +143,7 @@
 mod macho_signing;
 pub use macho_signing::*;
 mod macho_universal;
-pub use macho_universal::UniversalBinaryBuilder;
+pub use macho_universal::*;
 #[cfg(feature = "notarize")]
 pub mod notarization;
 #[cfg(feature = "notarize")]
diff -ruN -U3 a/src/macho.rs b/src/macho.rs
--- a/src/macho.rs	2025-12-09 17:09:33
+++ b/src/macho.rs	2025-12-09 15:10:07
@@ -646,9 +646,22 @@
         self.machos.iter()
     }
 
+    pub fn iter_macho_mut(&mut self) -> impl Iterator<Item = &mut MachOBinary<'a>> + '_ {
+        self.machos.iter_mut()
+    }
+
     pub fn nth_macho(&self, index: usize) -> Result<&MachOBinary<'a>, AppleCodesignError> {
         self.machos
             .get(index)
+            .ok_or(AppleCodesignError::InvalidMachOIndex(index))
+    }
+
+    pub fn nth_macho_mut(
+        &mut self,
+        index: usize,
+    ) -> Result<&mut MachOBinary<'a>, AppleCodesignError> {
+        self.machos
+            .get_mut(index)
             .ok_or(AppleCodesignError::InvalidMachOIndex(index))
     }
 }
diff -ruN -U3 a/src/signing.rs b/src/signing.rs
--- a/src/signing.rs	2025-12-09 17:09:33
+++ b/src/signing.rs	2025-12-10 22:12:19
@@ -66,6 +66,15 @@
         let output_path = output_path.as_ref();
 
         warn!("signing {} as a Mach-O binary", input_path.display());
+
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::PermissionsExt;
+            let mut perms = std::fs::metadata(input_path)?.permissions();
+            perms.set_mode(0o755);
+            std::fs::set_permissions(input_path, perms)?;
+        }
+
         let macho_data = std::fs::read(input_path)?;
 
         let mut settings = self.settings.clone();
